### PR TITLE
Picopass save written key

### PR DESCRIPTION
**Pull request by bettse**
# What's new

- When there is no user dictionary and the key of a card is changed to something the user input, save that value to a newly created user dictionary.  We believe we've seen people get into a bad state (key all 0's) when they don't know what they're doing.

# Verification 

- Backup/delete user dictionary
- Read a standard keyed iClass
- Write key (defaults to all 0's)
- See user dictionary with all 0's key was created
- Try to read card again, see it was successful using all 0's key from user dictionary.

# Checklist (For Reviewer)

- [x] PR has description of feature/bug or link to Confluence/Jira task
- [x] Description contains actions to verify feature/bugfix
- [x] I've built this code, uploaded it to the device and verified feature/bugfix

### CODE REVIEW

File: picopass/helpers/iclass_elite_dict.c [hunk 1]
```diff
@@ -147,7 +147,7 @@ bool iclass_elite_dict_add_key(IclassEliteDict* dict, uint8_t* key) {
     furi_assert(dict->stream);
 
     FuriString* key_str = furi_string_alloc();
-    for(size_t i = 0; i < 6; i++) {
+    for(size_t i = 0; i < ICLASS_ELITE_KEY_LEN; i++) {
```
Comment (skotopes) [inline picopass/helpers/iclass_elite_dict.c:150]: 8 is correct one?

File: picopass/scenes/picopass_scene_write_key.c [hunk 2]
```diff
@@ -17,6 +17,15 @@ NfcCommand picopass_scene_write_key_poller_callback(PicopassPollerEvent event, v
             event.data->req_write_key.key,
             picopass->write_key_context.key_to_write,
             PICOPASS_KEY_LEN);
+
+        // If there is no user dictionary, create it with the key they entered
+        // Prevent people who set all 0's from bricking their card
+        // TODO: Consider checking the elite user dict, when it exists, for the key
+        if(!iclass_elite_dict_check_presence(IclassEliteDictTypeUser)) {
+            IclassEliteDict* dict = iclass_elite_dict_alloc(IclassEliteDictTypeUser);
+            iclass_elite_dict_add_key(dict, picopass->write_key_context.key_to_write);
```
Comment (skotopes) [inline picopass/scenes/picopass_scene_write_key.c]:
```
(gdb) bt
#0  0x08011394 in __furi_crash_implementation () at furi/core/check.c:163
#1  0x08006636 in MemManage_Handler () at targets/f7/furi_hal/furi_hal_interrupt.c:283
#2  <signal handler called>
#3  iclass_elite_dict_add_key (dict=dict@entry=0x0, key=key@entry=0x2000ad50 "") at /Users/aku/Work/Flipper/flipperzero-good-faps/picopass/helpers/iclass_elite_dict.c:157
#4  0x2000e98c in picopass_scene_write_key_poller_callback (event=..., context=0x2000ac68) at /Users/aku/Work/Flipper/flipperzero-good-faps/picopass/scenes/picopass_scene_write_key.c:26
#5  0x2000c026 in picopass_poller_write_key_handler (instance=0x20018008) at /Users/aku/Work/Flipper/flipperzero-good-faps/picopass/protocol/picopass_poller.c:538
#6  0x2000bb86 in picopass_poller_callback (event=..., context=0x20018008) at /Users/aku/Work/Flipper/flipperzero-good-faps/picopass/protocol/picopass_poller.c:656
#7  0x08038a98 in nfc_iso14443a_poller_trx_custom_parity (instance=0x2000bb69 <picopass_poller_callback>, tx_buffer=0xa5a5a5a5, rx_buffer=0xa5a5a5a5, fwt=<optimized out>) at lib/nfc/nfc.c:446
#8  0x08038b1c in nfc_poller_trx (instance=0x0, tx_buffer=0x8, rx_buffer=0x20000284 <uxCriticalNesting>, fwt=536964632) at lib/nfc/nfc.c:480
#9  0xa5a5a5a4 in ?? ()
Backtrace stopped: previous frame identical to this frame (corrupt stack?)
(gdb) f 3 
#3  iclass_elite_dict_add_key (dict=dict@entry=0x0, key=key@entry=0x2000ad50 "") at /Users/aku/Work/Flipper/flipperzero-good-faps/picopass/helpers/iclass_elite_dict.c:157
157	        if(!stream_seek(dict->stream, 0, StreamOffsetFromEnd)) break;
```

**Review by skotopes — CHANGES_REQUESTED** [review]

File: picopass/scenes/picopass_scene_write_key.c (hunk 2 above)
Comment (skotopes) [reply picopass/scenes/picopass_scene_write_key.c]: was writing elite with all 00

File: picopass/helpers/iclass_elite_dict.c (hunk 1 above)
Comment (bettse) [reply picopass/helpers/iclass_elite_dict.c:150]: :100:

File: picopass/scenes/picopass_scene_write_key.c (hunk 2 above)
Comment (bettse) [reply picopass/scenes/picopass_scene_write_key.c]: > was writing elite with all 00

Expected behavior, but I'll look into the backtrace

Comment (bettse) [reply picopass/scenes/picopass_scene_write_key.c]: I moved where the write happens, I don't think it clicked with me before that it was doing the poller interactions, which is a bad place to put it.  I moved it to the on_enter and tried it a few times locally.

**Comment by skotopes** [comment]
```(gdb) bt
#0  0x080113a2 in __furi_crash_implementation () at furi/core/check.c:163
#1  0x08006636 in MemManage_Handler () at targets/f7/furi_hal/furi_hal_interrupt.c:283
#2  <signal handler called>
#3  iclass_elite_dict_add_key (dict=dict@entry=0x0, key=key@entry=0x2000a900 "") at /Users/aku/Work/Flipper/flipperzero-good-faps/picopass/helpers/iclass_elite_dict.c:157
#4  0x200108aa in picopass_scene_write_key_on_enter (context=0x2000a818) at /Users/aku/Work/Flipper/flipperzero-good-faps/picopass/scenes/picopass_scene_write_key.c:53
#5  0x2000f2d6 in picopass_scene_key_input_on_event (context=<optimized out>, event=...) at /Users/aku/Work/Flipper/flipperzero-good-faps/picopass/scenes/picopass_scene_key_input.c:35
#6  0x0806c848 in view_dispatcher_run (view_dispatcher=0x2000a290) at applications/services/gui/view_dispatcher.c:93
#7  0x200114bc in picopass_app (p=<optimized out>) at /Users/aku/Work/Flipper/flipperzero-good-faps/picopass/picopass.c:244
#8  0x0804be5c in flipper_application_thread (context=0x2000a460) at lib/flipper_application/flipper_application.c:224
#9  0x08013a74 in furi_thread_body (context=0x2000a740) at furi/core/thread.c:91
#10 0x08013a32 in furi_thread_catch () at furi/core/thread.c:62
```

Still crashing.

**Comment by skotopes** [comment]
/ext/apps_data/picopass folder is empty

**Comment by bettse** [comment]
I deleted the picopass folder and was able to reproduce.  I thought subfolders were handled automatically, but not a big deal to add them.  This should be good now, and I'll try to remember to test with no picopass for future stuff 👍

**Comment by bettse** [comment]
This is ready again and I think I fixed the crash

**Comment by skotopes** [comment]
Sorry for delay, I'm on vacation now and a little bit out of sync.

Key is not added to the dictionary.
Steps to reproduce:
- With dictionary exists on the sd card
- Write standard
- Write elite with custom key that is not in the dictionary
- Try to read, it fail
- Download dictionary and check that new key is not present

**Comment by bettse** [comment]
Ahh!  Have a good vacation: this can wait!


> With dictionary exists on the sd card

I assume you mean user dictionary, then this is expected.  This PR is only to cover the case when the user does not have a user dictionary: we’ve had some users who don’t know what they’re doing set the key to all zeros.  

If a user has a user dictionary then, for now, we assume they know what they are doing.

**Review by skotopes — APPROVED** [review]

**Comment by skotopes** [comment]
make sense, merged

**Pull request merged**